### PR TITLE
support failpoint nemesis

### PR DESCRIPTION
**Pull request by zyguan**
Signed-off-by: zyguan <zhongyangguan@gmail.com>

Here is an example:
```
$ jepsen_workload=append
$ jepsen_tarball=http://fileserver.pingcap.net/.../tidb-master-failpoint.tar.gz
$ jepsen_nodes=...
$ lein run test --time-limit=300 --concurrency 2n --version=latest --tarball-url=$jepsen_tarball --force-reinstall --username=root --ssh-private-key manifests/local/ssh-key --nodes=$jepsen_nodes --os centos --workload=$jepsen_workload --txn-mode=pessimistic --read-lock=update \
  --nemesis=failpoint --failpoints=tidb:tikvclient/twoPCRequestBatchSizeLimit:50%return,tidb:tikvclient/commitFailedSkipCleanup:30%return,tidb:tikvclient/prewritePrimaryFail:2%return,tidb:tikvclient/rpcFailOnRecv:2%return,tikv:delay_update_max_ts:10%return
...
2021-08-26 04:50:29,084{GMT}	INFO	[jepsen worker 4-172.32.1.5] jepsen.util: 4	:ok	:txn	[[:append 18 5]]	{:txn_scope "global", :start_ts 427282511655600155, :commit_ts 427282511707766800, :txn_commit_mode "1pc", :async_commit_fallback false, :one_pc_fallback false}
2021-08-26 04:50:29,085{GMT}	INFO	[jepsen worker 4-172.32.1.5] jepsen.util: 4	:invoke	:txn	[[:r 20 nil] [:r 18 nil] [:append 17 7] [:append 17 8]]
2021-08-26 04:50:29,129{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:enable-failpoint	(["172.32.1.1" ([(10080 "tikvclient/prewritePrimaryFail" "2%return") :ok] [(10080 "tikvclient/rpcFailOnRecv" "2%return") :ok])] ["172.32.1.4" ([(10080 "tikvclient/twoPCRequestBatchSizeLimit" "50%return") :ok] [(10080 "tikvclient/rpcFailOnRecv" "2%return") :ok])] ["172.32.1.2" ([(10080 "tikvclient/rpcFailOnRecv" "2%return") :ok] [(10080 "tikvclient/commitFailedSkipCleanup" "30%return") :ok] [(20180 "delay_update_max_ts" "10%return") :ok] [(10080 "tikvclient/prewritePrimaryFail" "2%return") :ok])] ["172.32.1.3" ([(20180 "delay_update_max_ts" "10%return") :ok] [(10080 "tikvclient/rpcFailOnRecv" "2%return") :ok] [(10080 "tikvclient/twoPCRequestBatchSizeLimit" "50%return") :ok] [(10080 "tikvclient/commitFailedSkipCleanup" "30%return") :ok])])
...
2021-08-26 04:50:38,865{GMT}	WARN	[jepsen worker 1-172.32.1.2] jepsen.core: Process 1 crashed
java.sql.SQLException: (conn=7) injected error on prewriting primary batch
...
2021-08-26 04:50:48,370{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:disable-failpoint	(["172.32.1.1" ([(10080 "tikvclient/prewritePrimaryFail") :ok] [(10080 "tikvclient/rpcFailOnRecv") :ok])] ["172.32.1.2" ([(10080 "tikvclient/commitFailedSkipCleanup") :ok] [(10080 "tikvclient/prewritePrimaryFail") :ok] [(10080 "tikvclient/rpcFailOnRecv") :ok] [(20180 "delay_update_max_ts") :ok])] ["172.32.1.3" ([(10080 "tikvclient/commitFailedSkipCleanup") :ok] [(10080 "tikvclient/rpcFailOnRecv") :ok] [(10080 "tikvclient/twoPCRequestBatchSizeLimit") :ok] [(20180 "delay_update_max_ts") :ok])] ["172.32.1.4" ([(10080 "tikvclient/rpcFailOnRecv") :ok] [(10080 "tikvclient/twoPCRequestBatchSizeLimit") :ok])] ["172.32.1.5" ()])
2021-08-26 04:50:48,383{GMT}	INFO	[jepsen worker 6-172.32.1.2] jepsen.util: 16	:ok	:txn	[[:r 61 [1 3]] [:append 62 1]]	{:txn_scope "global", :start_ts 427282516675395593, :commit_ts 427282516767145993, :txn_commit_mode "1pc", :async_commit_fallback false, :one_pc_fallback false}
2021-08-26 04:50:48,383{GMT}	INFO	[jepsen worker 6-172.32.1.2] jepsen.util: 16	:invoke	:txn	[[:r 60 nil]]
...
```

### CODE REVIEW

**Comment by cfzjywxk** [comment]
Maybe we need a way to connect  the failpoint implementation in the TiDB codes and other users including jepsen, so that we could quickly know what failpoints we could enable and if the enable has succeeded, like some sort of failpoint services?

**Comment by zyguan** [comment]
> Maybe we need a way to connect the failpoint implementation in the TiDB codes and other users including jepsen, so that we could quickly know what failpoints we could enable and if the enable has succeeded, like some sort of failpoint services?

There are two issues:
1. go failpoint doesn't collect failpoints declared in code, thus list api is unavailable.
2. client-go use `failpoint.Eval` directly, thus we cannot collect those failpoints (which are extremely useful) even when we solve the first issue.

Maybe we need an external tool to discover available failpoints and inject them to tidb-server.

**Comment by cfzjywxk** [comment]
> > Maybe we need a way to connect the failpoint implementation in the TiDB codes and other users including jepsen, so that we could quickly know what failpoints we could enable and if the enable has succeeded, like some sort of failpoint services?
> 
> There are two issues:
> 
> 1. go failpoint doesn't collect failpoints declared in code, thus list api is unavailable.
> 2. client-go use `failpoint.Eval` directly, thus we cannot collect those failpoints (which are extremely useful) even when we solve the first issue.
> 
> Maybe we need an external tool to discover available failpoints and inject them to tidb-server.

Any ideas for this failpoint discovery mechanism?

File: tidb/src/tidb/nemesis.clj [hunk 1]
```diff
@@ -88,6 +89,41 @@
...
+    (setup! [this test] this)
+
+    (invoke! [this test op]
+      (let [enable (fn [node]
+                     [node (doall
+                            (map #(try
+                                    (apply tu/fail-enable! node %)
+                                    [% :ok]
+                                    (catch Exception _
+                                      [% :fail]))
```
Comment (cfzjywxk) [inline tidb/src/tidb/nemesis.clj:104]: Will this print some logs in the result file?

File: tidb/src/tidb/core.clj [hunk 2]
```diff
@@ -363,6 +378,13 @@
        (map str/trim)
        (drop-while empty)))
 
+(defn parse-failpoints
```
Comment (cfzjywxk) [inline tidb/src/tidb/core.clj:381]: Do we need to add some unit tests for some commonly used failpoints in db or kv.

**Comment by zyguan** [comment]
> > > Maybe we need a way to connect the failpoint implementation in the TiDB codes and other users including jepsen, so that we could quickly know what failpoints we could enable and if the enable has succeeded, like some sort of failpoint services?
> > 
> > 
> > There are two issues:
> > 
> > 1. go failpoint doesn't collect failpoints declared in code, thus list api is unavailable.
> > 2. client-go use `failpoint.Eval` directly, thus we cannot collect those failpoints (which are extremely useful) even when we solve the first issue.
> > 
> > Maybe we need an external tool to discover available failpoints and inject them to tidb-server.
> 
> Any ideas for this failpoint discovery mechanism?

Yes, we can implement a tool, which scan the source code of tidb (as well as its dependencies) and rewrite the code (inject a http handler to status service) before compiling like failpoint-ctl.

File: tidb/src/tidb/nemesis.clj (hunk 1 above)
Comment (zyguan) [reply tidb/src/tidb/nemesis.clj:104]: Yes, just like `2021-08-26 04:50:29,129{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:enable-failpoint	(["172.32.1.1" ([(10080 "tikvclient/prewritePrimaryFail" "2%return") :ok] [(10080 "tikvclient/rpcFailOnRecv" "2%return") :ok])] ["172.32.1.4" ([(10080 "tikvclient/twoPCRequestBatchSizeLimit" "50%return") :ok] [(10080 "tikvclient/rpcFailOnRecv" "2%return") :ok])] ["172.32.1.2" ([(10080 "tikvclient/rpcFailOnRecv" "2%return") :ok] [(10080 "tikvclient/commitFailedSkipCleanup" "30%return") :ok] [(20180 "delay_update_max_ts" "10%return") :ok] [(10080 "tikvclient/prewritePrimaryFail" "2%return") :ok])] ["172.32.1.3" ([(20180 "delay_update_max_ts" "10%return") :ok] [(10080 "tikvclient/rpcFailOnRecv" "2%return") :ok] [(10080 "tikvclient/twoPCRequestBatchSizeLimit" "50%return") :ok] [(10080 "tikvclient/commitFailedSkipCleanup" "30%return") :ok])])`

File: tidb/src/tidb/core.clj (hunk 2 above)
Comment (zyguan) [reply tidb/src/tidb/core.clj:381]: To test the parse fn or to verify if db or kv supports those failpoints?

Comment (cfzjywxk) [reply tidb/src/tidb/core.clj:381]: Yes, seems their formats are a bit different.

Comment (zyguan) [reply tidb/src/tidb/core.clj:381]: They are actually same, here the format is `<tidb|tikv>:<failpoint_name>:<failpoint_value>[,...]`. The parse fn have some known cornor cases (eg, when value contains `,`), should we use a json formated string?

Comment (cfzjywxk) [reply tidb/src/tidb/core.clj:381]: As we may not use complex combinations by now, current format is ok.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Pull request merged**